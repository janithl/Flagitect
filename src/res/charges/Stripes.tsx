import React from 'react';
import { G, Path, Polygon, Svg } from 'react-native-svg';

import { coord } from '@lib/utils';

export enum StripedTypes {
  Serrated = 'Stripes Serrated',
  Wavy = 'Stripes Wavy',
}

export default (
  height: number,
  width: number,
  colour: string,
  stripedType = StripedTypes.Serrated,
  thickness = 20,
): JSX.Element => {
  thickness /= 100;

  const midpoint = height / 2;
  const t = thickness * midpoint;

  const renderStripes = () => {
    if (stripedType === StripedTypes.Serrated) {
      const points = [
        coord(0, midpoint - t),
        coord(midpoint * 0.5, midpoint - 1.5 * t),
        coord(midpoint, midpoint - t),
        coord(midpoint * 1.5, midpoint - 0.5 * t),
        coord(height, midpoint - t),
        coord(height, midpoint + t),
        coord(midpoint * 1.5, midpoint + 1.5 * t),
        coord(midpoint, midpoint + t),
        coord(midpoint * 0.5, midpoint + 0.5 * t),
        coord(0, midpoint + t),
      ];

      return <Polygon points={points.join(' ')} fill={colour} />;
    } else if (stripedType === StripedTypes.Wavy) {
      const points = [
        ['M', 0, midpoint - t].join(' '),
        ['A', midpoint, midpoint, '0 0 1', midpoint, midpoint - t].join(' '),
        ['A', midpoint, midpoint, '0 0 0', height, midpoint - t].join(' '),
        ['L', height, midpoint + t].join(' '),
        ['A', midpoint, midpoint, '0 0 1', midpoint, midpoint + t].join(' '),
        ['A', midpoint, midpoint, '0 0 0', 0, midpoint + t].join(' '),
        'Z',
      ];

      return <Path d={points.join(' ')} fill={colour} />;
    }

    return <G />;
  };

  const translateCoords = [0, Math.round((height - width) / 2)];
  return (
    <G transform={`translate(${translateCoords.join(' ')})`}>
      <Svg height={width} width={width} viewBox={`0 0 ${height} ${height}`}>
        {renderStripes()}
      </Svg>
    </G>
  );
};
