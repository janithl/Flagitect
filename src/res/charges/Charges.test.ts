import renderCharges, { ComplexTypes, SimpleTypes } from './index';
import { StripedTypes } from './Stripes';

describe('Charges', () => {
  it('renders Diamond correctly', () => {
    expect(
      renderCharges(
        [
          {
            id: 'diamond-1',
            type: SimpleTypes.Diamond,
            colour: '#c00',
            percentage: 50,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();
  });

  it('renders Pall correctly', () => {
    expect(
      renderCharges(
        [
          {
            id: 'pall-1',
            type: ComplexTypes.Pall,
            colour: '#0c0',
            thickness: 25,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();
  });

  it('renders Star correctly', () => {
    expect(
      renderCharges(
        [
          {
            id: 'star-5',
            type: ComplexTypes.Star,
            colour: '#0c0',
            thickness: 40,
            percentage: 50,
            points: 5,
            rotation: 90,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();

    expect(
      renderCharges(
        [
          {
            id: 'star-8',
            type: ComplexTypes.Star,
            colour: '#0c0',
            thickness: 60,
            percentage: 50,
            points: 8,
            rotation: 0,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();
  });

  it('renders Stripes correctly', () => {
    expect(
      renderCharges(
        [
          {
            id: 'stripes-serrated',
            type: StripedTypes.Serrated,
            colour: '#00c',
            thickness: 25,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();

    expect(
      renderCharges(
        [
          {
            id: 'stripes-wavy',
            type: StripedTypes.Wavy,
            colour: '#c00',
            thickness: 25,
          },
        ],
        400,
        600,
      ),
    ).toMatchSnapshot();
  });
});
