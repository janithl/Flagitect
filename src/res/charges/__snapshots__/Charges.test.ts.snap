// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charges renders Diamond correctly 1`] = `
<React.Fragment>
  <G
    id="charge-diamond-1"
  >
    <Polygon
      fill="#c00"
      points="150,200 300,100 450,200 300,300"
    />
  </G>
</React.Fragment>
`;

exports[`Charges renders Pall correctly 1`] = `
<React.Fragment>
  <G
    id="charge-pall-1"
  >
    <Polygon
      fill="#0c0"
      points="0,0 75,0 300,162.5 600,162.5 600,237.5 300,237.5 75,400 0,400 0,350 225,200 0,50"
    />
  </G>
</React.Fragment>
`;

exports[`Charges renders Star correctly 1`] = `
<React.Fragment>
  <G
    id="charge-star-5"
  >
    <G
      transform="rotate(90 300 200)"
    >
      <Polygon
        fill="#0c0"
        points="323.5,232.4 395.1,230.9 338,187.6 358.8,119.1 300,160 241.2,119.1 262,187.6 204.9,230.9 276.5,232.4 300,300"
      />
    </G>
  </G>
</React.Fragment>
`;

exports[`Charges renders Star correctly 2`] = `
<React.Fragment>
  <G
    id="charge-star-8"
  >
    <G
      transform="rotate(0 300 200)"
    >
      <Polygon
        fill="#0c0"
        points="323,255.4 370.7,270.7 355.4,223 400,200 355.4,177 370.7,129.3 323,144.6 300,100 277,144.6 229.3,129.3 244.6,177 200,200 244.6,223 229.3,270.7 277,255.4 300,300"
      />
    </G>
  </G>
</React.Fragment>
`;

exports[`Charges renders Stripes correctly 1`] = `
<React.Fragment>
  <G
    id="charge-stripes-serrated"
  >
    <G
      transform="translate(0 -100)"
    >
      <Svg
        height={600}
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 400 400"
        width={600}
      >
        <Polygon
          fill="#00c"
          points="0,150 100,125 200,150 300,175 400,150 400,250 300,275 200,250 100,225 0,250"
        />
      </Svg>
    </G>
  </G>
</React.Fragment>
`;

exports[`Charges renders Stripes correctly 2`] = `
<React.Fragment>
  <G
    id="charge-stripes-wavy"
  >
    <G
      transform="translate(0 -100)"
    >
      <Svg
        height={600}
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 400 400"
        width={600}
      >
        <Path
          d="M 0 150 A 200 200 0 0 1 200 150 A 200 200 0 0 0 400 150 L 400 250 A 200 200 0 0 1 200 250 A 200 200 0 0 0 0 250 Z"
          fill="#c00"
        />
      </Svg>
    </G>
  </G>
</React.Fragment>
`;
